cross_negative_script/cross_negative_script.gs

```diff
@@ -0,0 +1,216 @@
+                    with the option to use the original match type or convert all to exact match.
+    License:        https://github.com/we-discover/public/blob/master/LICENSE
+    Version:        1.0.0
+    Released:       2024-10-16
+    Author:         Nathan Ifill (@nathanifill)
+    Contact:        scripts@we-discover.com
+*/
+
+/************************************************* SETTINGS **************************************************************/
+
+// Options to choose between campaign-level and ad group-level negative keywords
+const ADD_CAMPAIGN_LEVEL_NEGATIVES = true;
+const ADD_AD_GROUP_LEVEL_NEGATIVES = true;
+const MAX_NEGATIVE_KEYWORDS = 5000; // This determines the maximum number of negative keywords to add to a campaign or ad group
+
+// Option to choose negative keyword match type
+const USE_ORIGINAL_MATCH_TYPE = false; // Set to false to use negative exact match for all or true to keep original match type
+
+// Option to filter which campaigns to include or exclude by name. If you leave this blank, the script will work across all
+// campaigns and/or ad groups. Leaving both of these blank is not recommended as (depending on the size of your account)
+// the script may not complete before Google's 30-minute execution limit.
+const INCLUDE_CAMPAIGN_NAMES = ['ZA > Search']; // Example: ['IncludePart1', 'IncludePart2']
+const EXCLUDE_CAMPAIGN_NAMES = []; // Example: ['ExcludePart1', 'ExcludePart2']
+
+// Option to enable or disable logging for faster execution
+const ENABLE_LOGGING = false; // Set to true to log what changes the script is making or set to false to speed up the script
+
+/*************************************************************************************************************************/
+
+function main() {
+  log("Starting process...");
+  log(" ");
+  log("Add campaign-level negatives? " + ADD_CAMPAIGN_LEVEL_NEGATIVES);
+  log("Add ad group-level negatives? " + ADD_AD_GROUP_LEVEL_NEGATIVES);
+  log("Use original match type for negatives? " + USE_ORIGINAL_MATCH_TYPE);
+  log("");
+  if (INCLUDE_CAMPAIGN_NAMES.length > 0) {
+    log("Campaign name must contain: " + INCLUDE_CAMPAIGN_NAMES);
+  }
+  if (EXCLUDE_CAMPAIGN_NAMES.length > 0) {
+    log("Campaign name must not contain: " + EXCLUDE_CAMPAIGN_NAMES);
+  }
+
+  if (ADD_CAMPAIGN_LEVEL_NEGATIVES) {
+    log("Adding campaign-level negatives:".toUpperCase());
+    log(" ");
+    addCampaignLevelNegatives();
+  }
+  if (ADD_AD_GROUP_LEVEL_NEGATIVES) {
+    log("Adding ad group-level negatives:".toUpperCase());
+    log(" ");
+    addAdGroupLevelNegatives();
+  }
+
+  log("Script processing complete. Have a nice day!");
+}
+
+/**
+ * Adds negative keywords at the campaign level.
+ */
+function addCampaignLevelNegatives() {
+  const campaigns = [];
+  const campaignIterator = AdsApp.campaigns().withCondition("Status = ENABLED").get();
+  while (campaignIterator.hasNext()) {
```

Comment:
🤔 I'm not sure if this is the most logical way of doing this -- with the need for three loops, two of which are nested.

Sure could be refactored to reduce the need for so many loops (which is always a good move), but as a slightly different approach, could you do something like:

- Grab all of the keywords from all campaigns that meet your filter using a GAQL query.
- Apply that in a single loop to those returned from your campaign iterator
- (If you combine the ad group and campaign loops) process ad groups within the campaign while you're there.

Reply:
I might be wrong here but I'm not sure that this would have the same functionality - particularly the bit about ad groups because of the and/or functionality of the script to do campaign-level and/or ad group level negatives. 

Unless I'm misunderstanding, what you've suggested would actually block the keywords in at least one campaign (if not all of them) because the GAQL query which is run once would grab all of the keywords in the account for campaign-level negatives. 

If you adjusted the GAQL query to exclude that particular campaign's keywords before applying the negatives, you'd have to make the same number of loops anyway, no? Maybe there could be a way to mass select everything and then check if each keyword is in the campaign/ad group its being applied to but that would mean even more operations than doing it the first way.

Reply:
I'm not close enough to the actual logic here - so taking your word for it